databricks_cli/dbfs/api.py

```diff
@@ -97,20 +107,72 @@ def file_exists(self, dbfs_path, headers=None):
             raise e
         return True
 
+    @retry(wait=wait_random_exponential(multiplier=1, max=30), 
+           retry=retry_if_exception_type(RateLimitException))
     def get_status(self, dbfs_path, headers=None):
-        json = self.client.get_status(dbfs_path.absolute_path, headers=headers)
-        return FileInfo.from_json(json)
+        try:
+            json = self.client.get_status(dbfs_path.absolute_path, headers=headers)
+            return FileInfo.from_json(json)
+        except HTTPError as e:
+            if e.response.status_code == 429:
+                click.echo("\rRate Limit Exceeded. Trying again.")
+                raise RateLimitException
+            raise e
+
+    @retry(wait=wait_random_exponential(multiplier=1, max=30), 
+           retry=retry_if_exception_type(RateLimitException))
+    def create_with_retries(self, dbfs_path, overwrite, headers):
+        try:
+            return self.client.create(dbfs_path, overwrite, headers=headers)
+        except HTTPError as e:
+            if e.response.status_code == 429:
+                click.echo("\rRate Limit Exceeded. Trying again.")
+                raise RateLimitException
+
+    @retry(wait=wait_random_exponential(multiplier=1, max=30), 
+           retry=retry_if_exception_type(RateLimitException))
+    def add_block_with_retries(self, handle, contents, headers):
+        try:
+            self.client.add_block(handle, b64encode(contents).decode(), headers=headers)
+        except HTTPError as e:
+            if e.response.status_code == 429:
+                click.echo("\rRate Limit Exceeded. Trying again.")
+                raise RateLimitException
+
+    @retry(wait=wait_random_exponential(multiplier=1, max=30), 
+           retry=retry_if_exception_type(RateLimitException))
+    def close_with_retries(self, handle, headers):
+        try:
+            self.client.close(handle, headers=headers)
+        except HTTPError as e:
+            if e.response.status_code == 429:
+                click.echo("\rRate Limit Exceeded. Trying again.")
+                raise RateLimitException
 
     def put_file(self, src_path, dbfs_path, overwrite, headers=None):
-        handle = self.client.create(dbfs_path.absolute_path, overwrite, headers=headers)['handle']
+        try:
+            handle = self.create_with_retries(dbfs_path, overwrite, headers=headers)['handle']
+        except HTTPError as e:
+            if e.response.json()['error_code'] == DbfsErrorCodes.RESOURCE_ALREADY_EXISTS:
+                pass
```

Comment:
What is the purpose of ignoring this exception? If it doesn't have anything to do with the 429 retries and it's a fix for something else then it should be on a separate PR.

Reply:
Ah old code which snuck in. My bad.